### DISCUSSION TITLE
Cancel github actions wokflow that uses if:always()

### DISCUSSION THREAD

**Discussion by hcortepi** [community/community#26303] [Actions] [answered]
<p>Hi there!</p>
<p>I am using github actions as the CI/CD tool to run automated test cases on a given application.</p>
<p>I have been struggling for a while with the <code>if:always()</code> tag, because I want to, at the end of the automation tests run, run a script in a test management platform to update it with the run results.</p>
<p>I want this to be done at all times, meaning that even when the step for the suite run fails, I want this step to happen to check the results. To achieve that I am forced to use the <code>if:always()</code> tag.</p>
<p>There is however some days where the plaftorm I use as test management tool is down or slow. When this happens this step that uses the always tag just keeps on running.</p>
<p>I’ve tried the cancelation thourgh the cancel workflow button or even runnning the command <code>gh run cancel &lt;run_id&gt;</code>, but nothing works.</p>
<p>Is it possible to have a force cancel workflow? If not, given I use a private repo, I will pay for extra minutes when I am not taking advantage of the tool.</p>
<p>Thanks in advance!<br>
Helena</p>

**Comment by jsoref** [comment, marked as answer]
<p>You might consider setting a timeout on your workflow.</p>
<p>This came up recently, and doesn’t appear to be well documented, so someone should file a bug.</p>
<p><a href="https://docs.github.com/en/actions/learn-github-actions/expressions#always" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/actions/learn-github-actions/expressions#always</a></p>
<p>In short, <code>always()</code> steps aren’t cancelable.</p>
<p>You might be able to use <code>success() || failure() || cancelled()</code> – I’m not sure if <code>cancelled()</code> has the same property.</p>

**Reply by ScottSuarez** [reply]
Why not just use if `if !cancelled()` in place of `always()` ?

Seems condensed logically IMO. 

`if success() || failure()` could also work assuming that `false` `false` is returned in case of cancelation.

`if success() || failure() || cancelled()` doesn't seem correct. This would still run in case of cancelation `false || false || true`. Thus I believe `if !cancelled()` to be the most condensed form of expression.

**Reply by TWiStErRob** [reply]
This is shown in the docs now: https://docs.github.com/en/actions/learn-github-actions/expressions#always

**Reply by kamilzki** [reply]
Please check force-cancel out: https://docs.github.com/en/rest/actions/workflow-runs#force-cancel-a-workflow-run

**Comment by GMNGeoffrey** [comment]
This is really weird behavior. If I cancel a job or a workflow I expect it to, you know, cancel. If I only cancel the jobs that that job depends on, it makes sense that it would run anyway, but if I do something to cancel that job itself, it is super unintuitive that it would not cancel

**Reply by fentas** [reply]
Well, imagine you need a cleanup step - e.g. shutting down infrastructure, etc. - even if you cancel, so you would define `always()` which runs always regardless of the state itself.

**Reply by fentas** [reply]
also, you can easily check if you want to ignore the `cancelled` state when you specify it more
`always() && needs.<job name before this one>.result != 'cancelled'`

**Reply by GMNGeoffrey** [reply]
Yes I understand the use case of a cleanup step or job. I'm referring to the case where I have cancelled *that* job, not its dependencies. The job does not cancel because it has `if: always()`. Checking `needs` doesn't help here because, again, I am cancelling the job itself.

**Reply by airtower-luna** [reply]
`if: '! cancelled()'` should do the trick: [`cancelled()`](https://docs.github.com/en/actions/learn-github-actions/expressions#cancelled) returns `true` if the job was cancelled, so the negation should result in "run unless cancelled".

**Reply by brianjmurrell** [reply]
I'm using:
```
    needs: previous-job
    if: ${{ always() && !contains('skip-previous-job', needs.Import-commit-message.outputs.message) }}
```

to create a workflow that allows a job that a later job *may* depend on to be skippable, without skipping the depending job.  I.e.:

A -> B

so that A runs before B, but A may be skipped and B can still function and should be run.  I have not been able to find a way to express that other than the above, which requires the `always()`, which then makes the job un-cancellable.

**Reply by ravwojdyla** [reply]
FYI this works best in my case:

```yaml
needs: previous-job
if: ${{ !cancelled() && (success() || failure() || needs.previous-job.result == 'skipped') }}
```

**Reply by GMNGeoffrey** [reply]
Yeah but unfortunately a "skipped" job shows up as a successful status for the purposes of required status checks

**Reply by jsoref** [reply]
Presumably you could add an extra step that looks to see if certain things are skipped and actively fails itself if it wanted to...

**Reply by GMNGeoffrey** [reply]
This already is that job, essentially. It's just looking at the status of other jobs and providing an aggregate status (so there's one top-level thing to look at for required checks and such). If I were to create something that checked it, how would I cancel that job? It is desirable to be able to cancel a workflow when GitHub runners are queuing a lot or unreachable (which happens disturbingly frequently).

**Comment by maximelebastard** [comment]
This `always()` tag seems to mean too much different things:
- Run it even if the previous task failed
- Run it even if the dependance is cancelled
- Run it even if my deployment is destroying everything and I just want to kill that deployment before its too late

This decision is hard to understand as a user

**Reply by TWiStErRob** [reply]
I think the infrastructure is doing what we ask it to do...
<details><summary><a href="https://www.google.com/search?q=define%3Aalways"><code>define:always</code></a></summary>

![image](https://github.com/community/community/assets/2906988/25a05a23-1f07-4ea8-b0c7-f875724e9c36)

</details>

I have a feeling that for all your cases there's a reasonable, more explicit alternative. e.g. for "Run it even if the previous task failed" if you don't put an `if: always()` on the step, then it won't execute if the previous one failed.

**Comment by XayOn** [comment]
So... once we hit this, how do we finally cancel a job that is *already* on that situation?
Our minutes are passing, and we can't cancel it!

**Reply by GMNGeoffrey** [reply]
![image](https://user-images.githubusercontent.com/5732088/214377625-2ca99fc3-5179-4cbd-b5c6-b132e247de5b.png)

**Reply by XayOn** [reply]
So... that sounds like a scam?

**Reply by jsoref** [reply]
Assuming you're writing a workflow, you can add various protections to it, a timeout, perhaps a background task that reads from some other state and `kill`s the main process if it sees your other state ...

**Reply by jtcressy** [reply]
so if I accidentally get into this scenario where I have an `always()` job that is stuck, it's stuck forever? I also have a concurrency rule so I'm not able to run any more jobs even after fixing the issue. 

Do I need to delete my repo and start over?

completely unacceptable!!!!

**Reply by jsoref** [reply]
If you don't like the behavior, open a ticket: https://support.github.com/

**Reply by PlaceReporter99** [reply]
> so if I accidentally get into this scenario where I have an always() job that is stuck, it's stuck forever? I also have a concurrency rule so I'm not able to run any more jobs even after fixing the issue.
>
> Do I need to delete my repo and start over?
>
> completely unacceptable!!!!

You might be able to change `cancel-in-progress` to `true` to cancel an action that is currently running when another action starts.

**Reply by TWiStErRob** [reply]
> so if I accidentally get into this scenario where I have an always() job that is stuck, it's stuck forever?

@jsoref the default timeout ["The maximum number of minutes to let a job run before GitHub automatically cancels it. Default: 360"](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes) doesn't apply here? i.e. does the cancellation not have a hard timeout?

Edit: my bad, the [hard deadline seems to be 6 hours](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits), and the wording there is "terminated" which I guess is stronger than "cancelled".

Edit: found one more [deadline for cancellation](https://docs.github.com/en/actions/managing-workflow-runs/canceling-a-workflow#steps-github-takes-to-cancel-a-workflow-run):
> After the 5 minutes cancellation timeout period, the server will force terminate all jobs and steps that don't finish running or fail to complete the cancellation process.

**Reply by dibas-indihood** [reply]
> > so if I accidentally get into this scenario where I have an always() job that is stuck, it's stuck forever?
> 
> @jsoref the default timeout ["The maximum number of minutes to let a job run before GitHub automatically cancels it. Default: 360"](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes) doesn't apply here? i.e. does the cancellation not have a hard timeout?
> 
> Edit: my bad, the [hard deadline seems to be 6 hours](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits), and the wording there is "terminated" which I guess is stronger than "cancelled".
> 
> Edit: found one more [deadline for cancellation](https://docs.github.com/en/actions/managing-workflow-runs/canceling-a-workflow#steps-github-takes-to-cancel-a-workflow-run):
> 
> > After the 5 minutes cancellation timeout period, the server will force terminate all jobs and steps that don't finish running or fail to complete the cancellation process.

I don't see the 5 minutes deadline is considered by GitHub, if there is timeout specified in the job. Can someone help me with timeout-minutes for job and workflow cancellation relationship ?

**Comment by Kolobok12309** [comment]
Some workaround for topic, you need use not `if: always()`, but `if: always() && !cancelled()`.
It successfully cancelled for my test job
```yml
custom_condition:
  if: always() && !cancelled()
  runs-on: ubuntu-latest
  steps:
    - run: echo "Hello world"
    - run: sleep 30
```

**Reply by GMNGeoffrey** [reply]
This workaround is mentioned above. Unfortunately it means that if a previous job was cancelled, this will return a status of "skipped" which counts as passing for required status checks. It doesn't really solve the issue.

**Reply by ChristopherHX** [reply]
_Just a crazy idea, I was also surprised to read skipped jobs are handled like success in branch protections in the docs_

I wonder what happens if you make GitHub throw an unhandled exception if and only if a previous job has been cancelled. Then there is only a small timeframe to make your concerns reality. (If you do reruns, I don't know what happens with a previously passing)
```yaml
custom_condition:
  if: always() && (contains(needs.*.result, 'cancelled') && fromjson('throw') || !cancelled())
  # needs in the job if has access to all transitive dependencies, but you may be surprised of the result field in matrix jobs 
  # always() && can be removed
  runs-on: ubuntu-latest
  needs: test
  steps:
    - run: echo "Hello world"
    - run: sleep 30
```
fromjson doesn't accept 'throw' as input and will cancel expression evaluation. (If this doesn't happen on initial evaluation you cannot cancel the job) 

_I even saw some weird skipped conclusions instead of a failure in the past_

**EDIT** Works flawless in my private repo
- pressing rerun deletes the previous passing or failed dependent job
- pressing cancel (before the custom_condition job) will make no job status appear in the commit status list
- pressing cancel during the custom_condition job will cancel the job

No idea if someone can trigger cancellation between job test (success) and job custom_condition (before if evaluation)

**EDIT 2**
- _The mentioned corner case above (e.g. pressing cancel after the previous job set it's result to success) seem to set conclusion to cancelled of the custom_condition job._ (see EDIT 3, why it isn't skipped)
- Other call trees of the same workflow continue without beeing affected by the `fromjson('throw')` code.

_This method is probably not officially supported, but it works if you want to abort execution of all dependent jobs without getting skipped in a PR_

**EDIT 3**
> This workaround is mentioned above. Unfortunately it means that if a previous job was cancelled, this will return a status of "skipped" which counts as passing for required status checks. It doesn't really solve the issue.

For me the conclusion is cancelled and not skipped, maybe GitHub changed behavior.

**Comment by Durinder** [comment]
For anyone stuck with an uncancellable run, you can use `gh run delete [<run-id>]` with the CLI: https://cli.github.com/manual/gh_run_delete

**Comment by rking-bennie** [comment]
How is no one here getting `unknown tag !<!cancelled()>` ? My github actions pipeline recognizes cancelled but if i attempt to ! it throws syntax error

**Reply by programmer04** [reply]
I got this too, it happens when `!` is the first character in the `if` expression, e.g.
- `if: !cancelled()` - error occurs
- `if: always() && !cancelled()` - works without problems

In YAML spec `!` is the non-specific tag, see https://yaml.org/spec/1.2.2/#332-resolved-tags.

But still, I think it should be considered a bug in GH actions, see related previous discussion https://github.com/github/docs/issues/7601#issuecomment-866684823 - they consider it as a corner case

**Reply by airtower-luna** [reply]
What if you explicitly make it a string, like `if: '!cancelled()'`? Haven't tried it, but logically it should be correct. :thinking:

**Reply by programmer04** [reply]
It works as expected @airtower-luna, thanks! But still, it's not a great experience...

Putting something in quotation is rather idiomatic when you want to evaluate it as a verbatim string than an expression. Furthermore, in VSCode it doesn't highlight and check the syntax for `if` stanza when quoted in opposite to unquoted.

I made PR to documentation https://github.com/github/docs/pull/27497

**Reply by ChristopherHX** [reply]
BTW `(!cancelled())` should also work, because `(` is not a reserved in yaml like `!`.

**Reply by brianjmurrell** [reply]
> What if you explicitly make it a string, like if: '!cancelled()'? Haven't tried it, but logically it should be correct. 🤔

Doesn't putting it in a string make it _truthy_ per the [warning](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution):

	 Using the ${{ }} expression syntax turns the contents into a string, and strings are truthy. For example, if: true && ${{ false }} will evaluate to true.

**Reply by programmer04** [reply]
I've tried `'!cancelled()'` and it works as expected (but it's not so obvious, because you rather expect something to be treated as a verbatim string when quoted in single quotes, not something that executes)

**Reply by airtower-luna** [reply]
@brianjmurrell, there are different levels here: One is YAML syntax, the other Actions expressions. The issue with `!cancelled()` (without the quotes) is that it's being interpreted as a *YAML* tag, and never makes it it the Actions parser. At the YAML level all Actions expressions are simply strings. :slightly_smiling_face: 

The issue you mention occurs if in the *Actions* evaluation the value of an `if` is a string in a way the author of the expression didn't expect.

**Comment by brianjmurrell** [comment]
I now seem to have an un-cancellable workflow that doesn't use `always()` in it anywhere.

**Comment by brianjmurrell** [comment]
Am I misunderstanding something or shouldn't `if: cancelled() || success() || failure()` be functionally the same as `if: always()`?  It seems they are not.  I could not get a job to run with the former until I added `always()` to it's condition with `if: always() || cancelled() || success() || failure()`.

I even put this `if: cancelled() || success() || failure()` into a job that proves that at least one of those is true and the job is still only run if I include the `always()` condition:

```yaml
  Debug-functional-if:
    name: Debug Functional if
    runs-on: [self-hosted, wolf]
    needs: [Build-RPM, Import-commit-message, Calc-rpm-build-matrix, Import-commit-pragmas,
            Import-commit-pragmas2]
    strategy:
      matrix:
        distro: ${{ fromJSON(needs.Calc-rpm-build-matrix.outputs.matrix) }}
      fail-fast: false
    # NOTE: the always() here is generally bad practice as it makes jobs uncancellable
    #       we only use it here because it's for debugging only
    if: always() || cancelled() || success() || failure()
    steps:
      - name: Is cancelled
        if: cancelled()
        run: echo "cancelled() == true"
      - name: Is success()
        if: success()
        run: echo "success() == true"
      - name: Is failure()
        if: failure()
        run: echo "failure() == true"
```

which reports:

![image](https://github.com/community/community/assets/559317/bd677350-6dcd-45a0-8f37-b6c452834077)

And only runs at all because I added the `always()` to the job's if:`.  If I remove it, the job is skipped.

**Reply by ChristopherHX** [reply]
> Am I misunderstanding something

Yeah for some legacy reason has success excluded skipped jobs (see azure pipelines docs, they describe it better only different naming)

`cancelled() || success() || failure() < = > always() && !skipped()`

`if: cancelled()` can not be cancelled either.

Better use `if: (!cancelled())`, run always as long not cancelled and get cancelled once pressing cancel

**Comment by brianjmurrell** [comment]
> Closed #26303 as resolved.

Is this really resolved?  I think the overall behaviour is quite silly and unintuitive and suspect there are still edge cases that are not achievable.

**Comment by tanin47** [comment]
I accidentally use `always()`. Now I'm unable to unstuck one of the runs.

`gh run delete <run-id>` returns:

```
HTTP 403: Could not delete the workflow run (URL)
```

What are my options here?

**Reply by ChristopherHX** [reply]
@tanin47 Did you try force cancel as well?

https://docs.github.com/en/rest/actions/workflow-runs#force-cancel-a-workflow-run

**Reply by tanin47** [reply]
Yes I did do that. It seems Github Actions is having an incident: https://www.githubstatus.com/. What a coincidence.

**Reply by njzjz** [reply]
> https://docs.github.com/en/rest/actions/workflow-runs#force-cancel-a-workflow-run

The force cancel API works for me. I think it should be selected as the answer to this discussion.
